compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -1759,14 +1775,52 @@ impl<'a> Parser<'a> {
             }
         }
 
-        let token = recovered.as_ref().unwrap_or(&self.token);
-        match Lit::from_token(token) {
+        // How many leading invisible delimiters are there of the right kind?
```

Comment:
If we want to parse a literal and see an invisible delimiter, then we probably just parse an expression and then report an error if the resulting expression is not a literal.

Not sure whether it should be done here in `fn parse_opt_lit` or 1-2 levels above.

Reply:
Looks like you are already doing that in `fn parse_literal_maybe_minus`.

Reply:
The tricky thing here is that `parse_unsuffixed_lit` cannot consume any tokens if it fails. Because `parse_meta_inner`, after such a failure, will cancel the error and then call `parse_meta_item`.

Reply:
In that case the disambiguation between `parse_unsuffixed_lit` and `parse_meta_item` needs to be done at `parse_unsuffixed_lit` level.

It can be done with look ahead 1, if it's a `Invisible(ExprMv)` then it's a literal, otherwise it's a meta-item.
The whole idea of using `err.cancel()` for syntax disambiguation is a bad practice.

Reply:
Good suggestion, done.